### PR TITLE
[VisBuilder-Next] Migrate legacy visualizations to visbuilder by mapping the url

### PR DESCRIPTION
**Pull request by SheyGao**
### Description

In this PR, we aim to migrate legacy visualization objects compatible with VisBuilder to VisBuilder objects, enabling users to directly open these legacy objects in VisBuilder. The method involves constructing a URL that navigates directly to the VisBuilder page rendering the object. Key changes to the existing codebase include:

1. Adding a new function, `visbuilderEditItem`, triggered when users click the `Import to VisBuilder` button. This function takes a legacy visualization item as input, migrating the states from the visualization object to those required in the VisBuilder URL by calling `constructVisBuilderPath`, and then navigates to the VisBuilder app with the constructed path to complete the migration.
2. Modifying the existing user interface in Visualize, allowing users to click `Edit`to open the object in its original format or `Import to VisBuilder` to open it in VisBuilder, instead of only having the `Edit` option previously.

### Issues Resolved
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7120

## Screenshot
1. Migrating an existing area-type legacy visualization object.

https://github.com/user-attachments/assets/3f00a22a-10d5-465f-a91b-ae34037f8b08 

2. Migrating a newly created line chart. 

https://github.com/user-attachments/assets/e49d0a1f-f74a-4f00-966f-317b32c77520

## Testing the changes

For testing purposes, we have added unit tests for two files: `edit_action_dropdown.test.tsx` to verify the updated UI changes, and `migrate_url.test.tsx` to validate the migration URL functionality. Additional functional tests will be included in a different repository.

## Changelog
- feat: [VisBuilder-Next] Migration of legacy visualizations to VisBuilder by constructing the URL. 

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/visualize/public/application/components/visualize_listing.tsx [hunk 1]
```diff
@@ -109,6 +110,16 @@ export const VisualizeListing = () => {
     [application, history]
   );
 
+  const { services: visualizeServices } = useOpenSearchDashboards<VisualizeServices>();
+
```
Comment (ananzh) [inline src/plugins/visualize/public/application/components/visualize_listing.tsx:114]: add a comment here

File: src/plugins/vis_builder/public/application/utils/get_top_nav_config.tsx [hunk 2]
```diff
@@ -228,7 +228,13 @@ export const getOnSave = (
             pathname: `${EDIT_PATH}/${id}`,
           });
         }
-        dispatch(setEditorState({ state: 'clean' }));
+        dispatch(setState({
+          editor: {
+            errors: {},
```
Comment (ananzh) [inline src/plugins/vis_builder/public/application/utils/get_top_nav_config.tsx:233]: should not overwrite errors

File: yarn.lock [hunk 3]
```diff
@@ -15986,7 +15986,7 @@ string-similarity@^4.0.1:
   resolved "https://registry.yarnpkg.com/string-similarity/-/string-similarity-4.0.4.tgz#42d01ab0b34660ea8a018da8f56a3309bb8b2a5b"
   integrity sha512-/q/8Q4Bl4ZKAPjj8WerIBJWALKkaPRfrvhfF8k/B23i4nzrlRj2/go1m90In7nG/3XDSbOo0+pu6RvCTM9RGMQ==
 
-"string-width-cjs@npm:string-width@^4.2.0", string-width@^4.1.0, string-width@^4.2.0, string-width@^4.2.3:
+"string-width-cjs@npm:string-width@^4.2.0":
```
Comment (ananzh) [inline yarn.lock:15989]: why we update yarn.lock? did we include any new package?

File: src/plugins/vis_builder/public/application/utils/get_top_nav_config.tsx (hunk 2 above)
Comment (ananzh) [reply src/plugins/vis_builder/public/application/utils/get_top_nav_config.tsx:233]:
```
dispatch(setState((prevState) => ({
  editor: {
    ...prevState.editor,
    errors: prevState.editor.errors ? { ...prevState.editor.errors } : {},
    state: 'clean',
  },
  isMigrated: false,
})));
```

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx [hunk 4]
```diff
@@ -413,10 +417,10 @@ class TableListView extends React.Component<TableListViewProps, TableListViewSta
   renderTable() {
     const selection = this.props.deleteItems
       ? {
-          onSelectionChange: (obj: Item[]) => {
+          onSelectionChange: (obj: Object[]) => {
```
Comment (ruanyl) [inline src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx:420]: Is this type used correctly? `Object[]`?

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx [hunk 5]
```diff
@@ -413,10 +417,10 @@ class TableListView extends React.Component<TableListViewProps, TableListViewSta
...
             this.setState({
               selectedIds: obj
-                .map((item) => item.id)
+                .map((item: any) => item.id)
```
Comment (ruanyl) [inline src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx:423]: Try to avoid `any` is possible.

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx [hunk 6]
```diff
@@ -440,10 +444,16 @@ class TableListView extends React.Component<TableListViewProps, TableListViewSta
...
+          <EditActionDropdown
+            item={item}
+            editItem={this.props.editItem}
+            visbuilderEditItem={this.props.visbuilderEditItem}
+          />
+        ),
       },
     ];
-
+
```
Comment (ruanyl) [inline src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx:456]: Cleanup the spaces?

File: src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts [hunk 7]
```diff
@@ -53,19 +55,19 @@ export const slice = createSlice({
...
+      (state.editor.errors ??= {})[key] = error;
     },
     setEditorState: (state, action: PayloadAction<{ state: EditorState }>) => {
       state.editor.state = action.payload.state;
     },
     setOriginatingApp: (state, action: PayloadAction<{ state?: string }>) => {
       state.originatingApp = action.payload.state;
     },
-    setState: (_state, action: PayloadAction<MetadataState>) => {
+    setState: (state, action: PayloadAction<MetadataState>) => {
```
Comment (joshuali925) [inline src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts:66]: why this change?

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        closePopover={closePopover}
+        panelPaddingSize="none"
+        initialFocus={false}
+      >
+        <EuiContextMenu initialPanelId={0} panels={panels} />
+      </EuiPopover>
+    );
+  };
+  
+  export default EditActionDropdown;
```
Comment (joshuali925) [inline src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx:86]: try to avoid default exports if no specific reason? i think most components don't use export default, unless they need to be lazy imported

File: src/plugins/vis_builder/public/application/utils/state_management/store.ts [hunk 9]
```diff
@@ -62,5 +62,5 @@ export type AppDispatch = Store['dispatch'];
 
 export { setState as setStyleState, StyleState } from './style_slice';
 export { setState as setVisualizationState, VisualizationState } from './visualization_slice';
-export { MetadataState } from './metadata_slice';
-export { setState as setUIStateState, UIStateState } from './ui_state_slice';
+export { setState as setMetadataState, MetadataState } from './metadata_slice';
+export { setState as setUIStateState, UIStateState } from './ui_state_slice'
```
Comment (ananzh) [inline src/plugins/vis_builder/public/application/utils/state_management/store.ts:66]: missed `;`

**Comment by ananzh** [comment]
Could you fix lint and rerun `yarn run lint:es && yarn run lint:style`? and also fix the unit test by running `yarn test:jest src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.test.tsx`

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+interface Item {
+    typeTitle: string;
+    id?: string;
+    version?: number;
+  }
+  
+  interface EditActionDropdownProps {
+    item: Item;
+    editItem?(item: object): void;
+    visbuilderEditItem?(item: object): void;
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            icon: 'pencil',
+            onClick: () => {
+              closePopover();
+              editItem && editItem(item);
+            },
+          },
+          ...(isSpecificType
+            ? [
+                {
+                  name: 'Import to VisBuilder',
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx [hunk 12]
```diff
@@ -57,14 +58,17 @@ interface Column {
 }
 
 interface Item {
+  typeTitle: string;
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx (hunk 6 above)
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    const panels = [
+      {
+        id: 0,
+        items: [
+          {
+            name: 'Edit',
+            icon: 'pencil',
+            onClick: () => {
+              closePopover();
+              editItem && editItem(item);
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx (hunk 8 above)
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx]: thanks, fixed

File: src/plugins/vis_builder/public/application/utils/state_management/store.ts (hunk 9 above)
Comment (SheyGao) [reply src/plugins/vis_builder/public/application/utils/state_management/store.ts]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+  
+    const closePopover = () => {
+      setPopoverOpen(false);
+    };
+  
+    // A saved object will only have the 'Import to VisBuilder' option 
+    // if it is a VisBuilder-compatible type and its version is 1.
+    const typeName = item.typeTitle;
+    const itemVersion = item.version;
+    const isSpecificType = types.includes(typeName) && itemVersion && itemVersion <= 1;
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx:44]: is this okay?

```
// A saved object will only have the 'Import to VisBuilder' option
// if it is a VisBuilder-compatible type and its version is <= 1.
  const typeName = item.typeTitle;
  const itemVersion = item.version;
  const isVisBuilderCompatible = types.includes(typeName) && itemVersion !== undefined && itemVersion <= 1;
```

File: src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/edit_action_dropdown.tsx:1]: thanks, fixed

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx (hunk 4 above)
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx]: thanks, changed it back to Item[]

File: src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx (hunk 5 above)
Comment (SheyGao) [reply src/plugins/opensearch_dashboards_react/public/table_list_view/table_list_view.tsx]: thanks, fixed

File: src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts [hunk 15]
```diff
@@ -53,19 +55,19 @@ export const slice = createSlice({
   reducers: {
     setError: (state, action: PayloadAction<{ key: string; error: boolean }>) => {
       const { key, error } = action.payload;
-      state.editor.errors[key] = error;
+      (state.editor.errors ??= {})[key] = error;
```
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts]: thanks, now `state.editor.errors` is defined

File: src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts (hunk 7 above)
Comment (SheyGao) [reply src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts:66]: thanks, changed it back

File: src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (SheyGao) [reply src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts]: If I changed it back to `state.editor.errors[key] = error;` in `src/plugins/vis_builder/public/application/utils/get_top_nav_config.tsx`, i have this


```
dispatch(
          setMetadataState({
            editor: {
              state: 'clean',
            },
            isMigrated: false,
          })
        );
```
        
and the error message is 
Property 'errors' is missing in type '{ state: "clean"; }' but required in type '{ errors: { [key: string]: boolean; }; state: EditorState; }'.ts(2741)
metadata_slice.ts(18, 5): 'errors' is declared here.
metadata_slice.ts(17, 3): The expected type comes from property 'editor' which is declared here on type 'MetadataState'

Is there anyway to not dispatch/override the errors and just keep the current errors?

[the remark this answers is not in the document]
Comment (joshuali925) [reply src/plugins/vis_builder/public/application/utils/state_management/metadata_slice.ts]: i think that was the suggestion here, to merge new state with the existing one https://github.com/opensearch-project/OpenSearch-Dashboards/pull/7529#discussion_r1699117832

**Review by ananzh — APPROVED** [review]

**Review by joshuali925 — APPROVED** [review]

**Review by LDrago27 — APPROVED** [review]

**Review by joshuali925 — APPROVED** [review]
general question for core team, with force pushes i find it hard to tell what the delta is comparing to the last time i reviewed. Is there a good way to see what changed?

doing a direct `..` comparison shows a lot more upstream changes through rebase rather than changes brought by this PR: https://github.com/opensearch-project/OpenSearch-Dashboards/compare/836974a163480915278a090fc0c955243b138d61..a5149689f199a2fabb19f02daf39674a7eea96d9
